In [4]:
!ls /sbgenomics/project-files

merged-tumor.bam  merged-tumor.bam.bai


In [41]:
import pysam
bamfile = pysam.AlignmentFile("/sbgenomics/project-files/merged-tumor.bam", "rb")

for read in bamfile:
    firstRead = read;
    break;
 
print("First read of BAM file:\n", firstRead)

bamfile.close()    

First read of BAM file:
 C0HVYACXX120402:7:1207:5722:57044	1187	20	9483248	27	76M	20	9483381	76	TTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAAAAAAAAACCAGAATGTGCCATTCTGATTTAAACTG	array('B', [28, 28, 27, 29, 31, 30, 31, 31, 29, 31, 35, 30, 29, 31, 34, 30, 29, 23, 41, 32, 20, 30, 29, 34, 34, 29, 30, 31, 30, 30, 30, 33, 33, 26, 39, 12, 25, 19, 32, 30, 35, 28, 35, 33, 23, 33, 35, 36, 30, 38, 33, 41, 34, 35, 31, 33, 23, 30, 30, 36, 27, 32, 29, 34, 35, 41, 33, 31, 33, 29, 32, 32, 31, 31, 31, 34])	[('XA', 'GL000217.1,-110754,76M,1;'), ('BD', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('MD', '76'), ('RG', '1'), ('BI', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('NM', 0), ('MQ', 27), ('AS', 76), ('XS', 71)]


Flag field is (value 1187) is decoded using https://broadinstitute.github.io/picard/explain-flags.html:

    read paired (0x1)
    read mapped in proper pair (0x2)
    mate reverse strand (0x20)
    second in pair (0x80)
    read is PCR or optical duplicate (0x400)

In [42]:
import pysam
bamfile = pysam.AlignmentFile("/sbgenomics/project-files/merged-tumor.bam", "rb")

print("Number of unmapped reads in BAM file: ", bamfile.unmapped)

bamfile.close()

Number of unmapped reads in BAM file:  0


In [28]:
import pysam
bamfile = pysam.AlignmentFile("/sbgenomics/project-files/merged-tumor.bam", "rb")

num_of_reads = 0
for read in bamfile:
    num_of_reads+=1;
print("Number of reads in BAM file: ", num_of_reads)

bamfile.close()

Number of reads in BAM file:  2921629


In [44]:
import pysam
bamfile = pysam.AlignmentFile("/sbgenomics/project-files/merged-tumor.bam", "rb")

map_qual_zero_num = 0
for read in bamfile:
    if (read.mapping_quality == 0):
        map_qual_zero_num+=1;
print("Number of reads with mapping quality 0: ", map_qual_zero_num)

bamfile.close()

Number of reads with mapping quality 0:  126628


In [39]:
import pysam
bamfile = pysam.AlignmentFile("/sbgenomics/project-files/merged-tumor.bam", "rb")

num = 0
sum_of_qual = 0
avg = 0

for read in bamfile:
    num+=1;
    sum_of_qual += read.mapping_quality;

avg = sum_of_qual/num    
print("Average mapping quality: ", avg)

bamfile.close()

Average mapping quality:  55.91379158681681


In [38]:
import pysam
bamfile = pysam.AlignmentFile("/sbgenomics/project-files/merged-tumor.bam", "rb")

num = 0
sum_of_qual = 0
avg = 0

for read in bamfile:
    if(read.mapping_quality != 0):
        num+=1;
        sum_of_qual += read.mapping_quality;

avg = sum_of_qual/num    
print("Average mapping quality of reads with non-zero mapping quality: ", avg)

bamfile.close()

Average mapping quality of reads with non-zero mapping quality:  58.446975510921106
